src/System.Text.Http/System/Text/Http/Parser/HttpParser.cs

```diff
@@ -44,714 +41,310 @@ public HttpParser(bool showErrorDetails)
-                    return false;
-                }
-            }
 
-            // Fix and parse the span
-            fixed (byte* data = &MemoryMarshal.GetReference(span))
+            if (!reader.IsNext(Eol, advancePast: true))
             {
-                ParseRequestLine(ref handler, data, span.Length);
+                // Not CR/LF
+                RejectRequestLine(requestLine);
             }
 
-            examined = consumed;
+            // Parse the span
+            ParseRequestLine(handler, requestLine);
             return true;
         }
 
-        public unsafe bool ParseRequestLine<T>(ref T handler, in ReadOnlySequence<byte> buffer, out int consumed) where T : IHttpRequestLineHandler
+        public bool ParseRequestLine<T>(T handler, in ReadOnlySequence<byte> buffer, out SequencePosition consumed, out SequencePosition examined) where T : IHttpRequestLineHandler
         {
-            // Prepare the first span
-            var span = buffer.First.Span;
-            var lineIndex = span.IndexOf(ByteLF);
-            if (lineIndex >= 0)
+            consumed = buffer.Start;
+            examined = buffer.End;
+            var reader = new BufferReader<byte>(buffer);
+            bool success = ParseRequestLine(handler, ref reader);
+            consumed = reader.Position;
+            if (success) examined = reader.Position;
+            return success;
+        }
+
+        public bool ParseRequestLine<T>(T handler, in ReadOnlySequence<byte> buffer, out int consumed) where T : IHttpRequestLineHandler
+        {
+            // Look for CR/LF
+
+            BufferReader<byte> reader = new BufferReader<byte>(buffer);
+            if (!reader.TryReadToAny(out ReadOnlySpan<byte> requestLine, Eol, advancePastDelimiter: false))
             {
-                consumed = lineIndex + 1;
-                span = span.Slice(0, lineIndex + 1);
+                // Couldn't find a delimiter
+                consumed = 0;
+                return false;
             }
-            else
+
+            if (!reader.IsNext(Eol, advancePast: true))
             {
-                var eol = Sequence.IndexOf(buffer, s_Eol[0], s_Eol[1]);
-                if (eol == -1)
+                if (reader.TryRead(out byte value) && value == ByteCR && !reader.TryRead(out value))
                 {
+                    // Incomplete if ends in CR
                     consumed = 0;
                     return false;
                 }
 
-                span = PipelineExtensions.ToSpan(buffer.Slice(0, eol + s_Eol.Length));
+                // Not CR/LF
+                RejectRequestLine(requestLine);
             }
 
-            // Fix and parse the span
-            fixed (byte* data = &MemoryMarshal.GetReference(span))
-            {
-                ParseRequestLine(ref handler, data, span.Length);
-            }
+            ParseRequestLine(handler, requestLine);
 
-            consumed = span.Length;
+            consumed = (int)reader.Consumed;
```

Comment:
Can you please add a comment here explaining why this cast is safe here?

Reply:
It isn't really, but we weren't checking for it before.

@pakrym, any thoughts on this in general? I can turn this into a checked cast, but it would be good to understand what our expectations are with >2GB headers.